python/cudf/cudf/core/frame.py

```diff
@@ -730,8 +730,8 @@ def where(self, cond, other=None, inplace=False):
             if len(common_cols) > 0:
                 # If `self` and `cond` are having unequal index,
                 # then re-index `cond`.
-                if len(self.index) != len(cond.index) or any(
-                    self.index != cond.index
+                if len(self.index) != len(cond.index) or not self.index.equals(
```

Comment:
`Index.equals` will always return False if the lengths aren't equal (we short circuit) so I think we can remove the length check from here.